### DISCUSSION TITLE
How Prometheus manages hash collisions  on SeriesID

### DISCUSSION THREAD

**Discussion by evanxg852000** [prometheus/prometheus#11942] [General]
So I have noticed Prometheus generates a `SerieID` upon receiving a data scraped. it combines the label set to generate a unique id from the hash. But I am concerned about the hash collisions. That is two different metrics with probably two different label set could hash to the same serieID. Does Prometheus puts these data point from different series in the same series or how does it handle it?

**Comment by bboreham** [comment]
I think you are asking about `Head.lastSeriesID`: https://github.com/prometheus/prometheus/blob/a70d474fdfde5fdb40f8900c80ee74249b6bc245/tsdb/head.go#L76

This is not a hash; it is incremented each time a reference is given out, so they are unique for one Prometheus until 2^64 series have been created.

Where hashes are used to look up series in the Head, care is taken to find the exact right one:
https://github.com/prometheus/prometheus/blob/a70d474fdfde5fdb40f8900c80ee74249b6bc245/tsdb/head.go#L1682-L1683

However there are issues with hash collisions in other contexts; see #12519.
